### PR TITLE
Update Predictive Cache configuration and implementation

### PR DESCRIPTION
**Pull request by kried**
### Description

Predictive cache feature parity with Android , [#6071](https://github.com/mapbox/mapbox-navigation-android/pull/6071), and [#5927](https://github.com/mapbox/mapbox-navigation-android/pull/5927).
Takes advantage of the results from [#3916](https://github.com/mapbox/mapbox-navigation-native/pull/3916)

### Implementation
Includes fixes:
1. Implemented map predictive cache with `TilesetDescriptor` so that volatile sources are not loaded unexpectedly. In Android [#5068](https://github.com/mapbox/mapbox-navigation-android/pull/5068), in NN [#3916](https://github.com/mapbox/mapbox-navigation-native/pull/3916)
2. Added separate configuration for map and navigation options.
3. Added caching of tilesets that were loaded after an instance of PredictiveCacheManager was created.
4. Added zoomRange for maps cache configurations.

~~In this pull request `minZoom` and `maxZoom` are not added  to the `PredictiveCacheMapsOptions` because they can be used only with `TilesetDescriptorOptions`, member  of `MapboxMaps` in iOS. And because `PredictiveCacheManager` is part of `MapboxCoreNavigation` it cannot depend on `MapboxMaps`~~

In this pull request was not added the usage `maximumConcurrentRequests`, because it is not allowed by NN now. Updates are needed in Nav Native for passing `maximumConcurrentRequests` when creating `PredictiveCacheController` with `TilesetDescriptor`

### Questions
Deprecation of old properties is questionable. I would be glad for your feedback, is it worth removing the deprecation

### CODE REVIEW

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift [hunk 1]
```diff
@@ -42,19 +47,31 @@ public class PredictiveCacheManager {
...
     }
     
     deinit {
         unsubscribeNotifications()
+        observer?.cancel()
+    }
+
+    private func subscribeToStyleLoadEvents() {
+        self.observer = map?.onEvery(event: .styleLoaded) { [weak self] _ in
+            self?.updateControllers()
```
Comment (kried) [inline Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift:73]: Parity with [Android](https://github.com/mapbox/mapbox-navigation-android/blob/f28c864c61c576624c1682062b99798bae30f690/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt#L180)

```(kotlin)
        val onStyleLoadedListener = OnStyleLoadedListener {
        ...
        updateMapsControllers(
```
Allows to cache styles that was not loaded at the moment of PredictiveCacheManager creation

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheOptions.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     Predictive cache Map related options
+     */
+    public var predictiveCacheMapsOptions: PredictiveCacheMapsOptions = .init()
+    
+    /**
+     How far around the user's location caching is going to be performed.
+     
+     Defaults to 2000 meters.
+     */
+    @available(*, deprecated, message: "Use `predictiveCacheNavigationOptions` and `predictiveCacheMapsOptions` instead.")
```
Comment (kried) [inline Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheOptions.swift:25]: Not sure, maybe it should not be marked as deprecated in order to simplify the case when users need the same options for maps and navigation

File: Sources/MapboxNavigation/NavigationMapView.swift [hunk 3]
```diff
@@ -2302,10 +2304,20 @@ open class NavigationMapView: UIView {
...
-        let styleSourcePaths = mapView.styleSourceDatasets(["raster", "vector"])
-        
-        predictiveCacheManager = PredictiveCacheManager(predictiveCacheOptions: predictiveCacheOptions,
-                                                        styleSourcePaths: styleSourcePaths)
+        predictiveCacheMapObserver?.cancel()
+
+        self.predictiveCacheManager = PredictiveCacheManager(predictiveCacheOptions: predictiveCacheOptions,
+                                                             styleSourcePaths: styleSourceDatasetsForCaching())
+
+        self.predictiveCacheMapObserver = mapView.mapboxMap?.onEvery(event: .styleLoaded) { [weak self] _ in
```
Comment (kried) [inline Sources/MapboxNavigation/NavigationMapView.swift:2312]: Parity with [Android](https://github.com/mapbox/mapbox-navigation-android/blob/main/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt#L229)

```kotlin
        val onStyleLoadedListener = OnStyleLoadedListener {
        ...
            createMapsController(styleURI, map, tileStore)
```
Allows to cache styles that was not loaded at the moment of PredictiveCacheManager creation.

The observer is stored to correctly handle the case when method `enablePredictiveCaching` is called multiple times

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        let manager = PredictiveCacheManager(predictiveCacheOptions: predictiveCacheOptions,
+                                             mapOptions: (tileStore: tilestore, styleSourcePaths: ["first"]),
+                                             navigator: navigator)
+        let expectedOptions = predictiveCacheOptions.predictiveCacheNavigationOptions.locationOptions
+        XCTAssertEqual(navigator.passedNavigationTrackerOptions?.currentLocationRadius, UInt32(expectedOptions.currentLocationRadius))
+        XCTAssertEqual(navigator.passedNavigationTrackerOptions?.destinationLocationRadius, UInt32(expectedOptions.destinationLocationRadius))
+        XCTAssertEqual(navigator.passedNavigationTrackerOptions?.routeBufferRadius, UInt32(expectedOptions.routeBufferRadius))
+
+        let expectedMapOptions = predictiveCacheOptions.predictiveCacheMapsOptions.locationOptions
+        XCTAssertTrue(navigator.passedTileStore == tilestore)
```
Comment (Udumft) [inline Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift:51]: NIT: `XCTAssertEqual`

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        XCTAssertEqual(navigator.passedCacheOptions?.dataset, "first")
+        XCTAssertEqual(navigator.passedCacheOptions?.dataDomain, .maps)
+        XCTAssertEqual(navigator.passedCacheOptions?.concurrency, predictiveCacheOptions.predictiveCacheMapsOptions.maximumConcurrentRequests)
+
+        navigator.passedLocationTrackerOptions = nil
+        navigator.passedNavigationTrackerOptions = nil
+
+        manager.updateMapControllers(styleSourcePaths: ["second"])
+
+        XCTAssertNil(navigator.passedNavigationTrackerOptions, "Does not recreate navigation controller")
```
Comment (Udumft) [inline Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift:65]: This basically repeats `testCreateNavigationController`

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift [hunk 6]
```diff
@@ -0,0 +1,93 @@
+import XCTest
+import MapboxCommon
+import MapboxNavigationNative
+@testable import TestHelper
+@testable import MapboxCoreNavigation
+
+final class PredictiveCacheManagerTests: TestCase {
+    var predictiveCacheOptions = PredictiveCacheOptions()
+    let navigator = NativeNavigatorSpy()
```
Comment (Udumft) [inline Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift:9]: These instances are shared between tests which compromises tests isolation. Consider recreating them in `setUp` or provide computed getters.

**Review by Udumft — COMMENTED** [review]
`TilesetDescriptorOptions` are a member of `Maps`, but we don't need that type in `Core`, we need `TilesetDescriptor` which is a member of `Common` and is available in `Core`. What if we add `PredictiveCacheManager.updateMapControllers` version with `TilesetDescriptor`s argument, but do `TilesetDescriptorOptions -> TilesetDescriptor` transformation at `NavigationMapView` (probably even as `PredictiveCacheManager` extension).

And also, please do not forget to include new and moved files into `xcodeproj`.

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheOptions.swift (hunk 2 above)
Comment (Udumft) [reply Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheOptions.swift:25]: I think that deprecation makes sense here because old implementation assumes that both values are the same, but if it's not, `getter` will become incorrect. And it is also a good practice to point developers to use recommended setup.

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift (hunk 5 above)
Comment (kried) [reply Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift]: I removed this line, but left checking of map cache controller creation

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift (hunk 6 above)
Comment (kried) [reply Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift]: I moved creation of spies to setUp

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift (hunk 4 above)
Comment (kried) [reply Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheManagerTests.swift]: Fixed

**Comment by kried** [comment]
Implemented map caching through `TilesetDescriptor`.  Filtering is done completely on the side of Nav Native, as described mapbox/mapbox-navigation-android#4707
Public interface of `PredictiveCacheManager` is preserved, but deprecated `styleSourcePaths` ignored.

File: MapboxNavigation.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -586,6 +591,17 @@
...
+			isa = PBXCopyFilesBuildPhase;
+			buildActionMask = 2147483647;
+			dstPath = "";
+			dstSubfolderSpec = 10;
+			files = (
+				2C67751D290AC4A6009C5EE4 /* MapboxNavigationNative.xcframework in Embed Frameworks */,
+			);
+			name = "Embed Frameworks";
+			runOnlyForDeploymentPostprocessing = 0;
+		};
```
Comment (Udumft) [inline MapboxNavigation.xcodeproj/project.pbxproj:604]: Did you intentionally changed that?

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheMapsOptions.swift [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public var maximumConcurrentRequests: UInt32 = 2
+
+    /**
+     Closed range zoom level for the tile package..
+     See `TilesetDescriptorOptionsForTilesets.minZoom` and `TilesetDescriptorOptionsForTilesets.maxZoom`.
+
+     Defaults to 0..16.
+     */
+    public var zoomRange: ClosedRange<UInt8> = 0...16
```
Comment (Udumft) [inline Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheMapsOptions.swift:26]: I think we should also mention in CHANGELOG that this setting became available

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+      */
+    public convenience init(predictiveCacheOptions: PredictiveCacheOptions,
+                            cacheMapOptions: CacheMapOptions? = nil) {
+        self.init(predictiveCacheOptions: predictiveCacheOptions,
+                  cacheMapOptions: cacheMapOptions,
+                  navigatorProvider: Navigator.self)
+    }
+
+    /**
+     IInitializes a predictive cache.
```
Comment (Udumft) [inline Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift:50]: Typo 'IInitializes'

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        var mapOptions: MapOptions?
+        if let tileStore = NavigationSettings.shared.tileStoreConfiguration.mapLocation?.tileStore {
+            mapOptions = MapOptions(tileStore, styleSourcePaths)
+        }
+
+        self.init(predictiveCacheOptions: predictiveCacheOptions, mapOptions: mapOptions)
+    }
+    
+    /**
+     IInitializes a predictive cache.
```
Comment (Udumft) [inline Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheManager.swift:69]: Same typo

**Review by Udumft — APPROVED** [review]
LGTM with few NITS

File: Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheMapsOptions.swift (hunk 8 above)
Comment (kried) [reply Sources/MapboxCoreNavigation/PredictiveCache/PredictiveCacheMapsOptions.swift:26]: Updated changelog

File: MapboxNavigation.xcodeproj/project.pbxproj (hunk 7 above)
Comment (kried) [reply MapboxNavigation.xcodeproj/project.pbxproj]: I rolled it back. 

It was an attempt to fix the test run from `MapboxNavigation.xcodeproj`, but the problem with them was the lack of support for Carthage for `MapboxNavigation` target

File: MapboxNavigation.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -1290,6 +1305,7 @@
 				E2C1CDB2265E439300E158E0 /* MapboxMobileEvents.xcframework in Frameworks */,
 				E2C1CDB5265E43D500E158E0 /* Quick.xcframework in Frameworks */,
 				E2C1CDB7265E43E100E158E0 /* Nimble.xcframework in Frameworks */,
+				2C67751C290AC4A6009C5EE4 /* MapboxNavigationNative.xcframework in Frameworks */,
```
Comment (MaximAlien) [inline MapboxNavigation.xcodeproj/project.pbxproj:1308]: Nit: is this one really needed? Just want to make sure we do not add any excessive dependency.

Comment (kried) [reply MapboxNavigation.xcodeproj/project.pbxproj:1310]: This new one is needed for test for my new `NativeNavigatorSpy` in TestHelper. It adds a dependency on `NavigationNative` for TestHelper.
Without it we got build error such as 
> error build: Undefined symbol: _OBJC_CLASS_$_MBNNNavigator

File: Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheOptionsTests.swift [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import CoreLocation
+import MapboxCoreNavigation
+import XCTest
+@testable import MapboxDirections
```
Comment (Udumft) [inline Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheOptionsTests.swift:4]: Why is this import needed?

Comment (kried) [reply Tests/MapboxCoreNavigationTests/PredictiveCache/PredictiveCacheOptionsTests.swift]: Deleted, there's no need to import

**Comment by kried** [comment]
Tests
`MapboxNavigationTests:
		VanishingRouteLineTests.testEmptyRouteGradientStops()
		VanishingRouteLineTests.testEmptyRouteWithValidRouteProgress()
`
 also fail on [main](https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-ios/9998/workflows/d3aa47b6-dd46-46d9-aca7-1fef7f0de825/jobs/92650)
